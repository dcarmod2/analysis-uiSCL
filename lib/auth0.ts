import {initAuth0} from '@auth0/nextjs-auth0'
import {ISignInWithAuth0} from '@auth0/nextjs-auth0/dist/instance'
import {ISession} from '@auth0/nextjs-auth0/dist/session/session'
import {parse} from 'cookie'
import {IncomingMessage} from 'http'
import ms from 'ms'

import {IUser} from './user'

const cookieLifetime = ms('30 days') / 1000
const httpTimeout = ms('10s')
const scope = 'openid profile id_token'

/**
 * Auth0 is initialized with the origin taken from an incoming header. With Vercel Preview Deployments,
 * a single lambda may handle multiple origins. Ex: conveyal.dev, branch-git.conveyal.vercel.app,
 * and analysis-11234234.vercel.app. Therefore we initialize a specific auth0 instance for each origin
 * and store them based on that origin.
 */
const auth0s = {
  // [origin]: ISignInWithAuth0
}

function createAuth0(origin: string): ISignInWithAuth0 {
  if (process.env.NEXT_PUBLIC_AUTH_DISABLED === 'true') {
    return {
      handleCallback: async () => {},
      handleLogin: async () => {},
      handleLogout: async () => {},
      handleProfile: async () => {},
      getSession: async (): Promise<ISession> => ({
        createdAt: Date.now(),
        user: {
          name: 'local',
          'http://conveyal/accessGroup': 'local'
        },
        idToken: 'fake'
      }),
      requireAuthentication: (fn) => fn,
      tokenCache: () => ({
        getAccessToken: async () => ({})
      })
    }
  } else {
    return initAuth0({
      clientId: process.env.AUTH0_CLIENT_ID,
      clientSecret: process.env.AUTH0_CLIENT_SECRET,
      scope,
      domain: process.env.AUTH0_DOMAIN,
      redirectUri: `${origin}/api/callback`,
      postLogoutRedirectUri: origin,
      session: {
        cookieSecret: process.env.SESSION_COOKIE_SECRET,
        cookieLifetime,
        storeIdToken: true
      },
      oidcClient: {
        httpTimeout
      }
    })
  }
}

// Dyanmically create the Auth0 instance based upon a request
export default function initAuth0WithReq(
  req: IncomingMessage
): ISignInWithAuth0 {
  const host = req.headers.host
  const protocol = /^localhost(:\d+)?$/.test(host) ? 'http:' : 'https:'
  const origin = `${protocol}//${host}`
  if (auth0s[origin]) return auth0s[origin]
  auth0s[origin] = createAuth0(origin)
  return auth0s[origin]
}

/**
 * Flatten the session object and assign the accessGroup without the http portion.
 */
export async function getUser(req: IncomingMessage): Promise<IUser> {
  const auth0 = initAuth0WithReq(req)
  const session = await auth0.getSession(req)
  if (!session) {
    throw new Error('User session does not exist. User must be logged in.')
  }

  const user = {
    // This is a namespace for a custom claim. Not a URL: https://auth0.com/docs/tokens/guides/create-namespaced-custom-claims
    accessGroup: session.user['http://conveyal/accessGroup'],
    adminTempAccessGroup: null,
    email: session.user.name,
    idToken: session.idToken
  }

  if (user.accessGroup === process.env.NEXT_PUBLIC_ADMIN_ACCESS_GROUP) {
    const adminTempAccessGroup = parse(req.headers.cookie || '')
      .adminTempAccessGroup
    if (adminTempAccessGroup) user.adminTempAccessGroup = adminTempAccessGroup
  }

  return user
}

/**
 * Helper function for retrieving the access group.
 */
export async function getAccessGroup(req: IncomingMessage): Promise<string> {
  const user = await getUser(req)
  if (user.adminTempAccessGroup && user.adminTempAccessGroup.length > 0) {
    return user.adminTempAccessGroup
  }
  return user.accessGroup
}
