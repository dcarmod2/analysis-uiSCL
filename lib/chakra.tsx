import {CSSReset, ThemeProvider, theme, ITheme} from '@chakra-ui/core'
import React from 'react'

const ConveyalTheme: ITheme = {
  ...theme,
  fonts: {
    ...theme.fonts,
    body: `'Inter', -apple-system, 'Segoe UI', sans-serif`
  },
  fontWeights: {
    ...theme.fontWeights,
    normal: 400,
    medium: 400,
    bold: 600
  }
}

export default function ChakraTheme({children}) {
  return (
    <ThemeProvider theme={ConveyalTheme}>
      <CSSReset />
      {children}
    </ThemeProvider>
  )
}
