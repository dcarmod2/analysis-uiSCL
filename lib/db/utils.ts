import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
