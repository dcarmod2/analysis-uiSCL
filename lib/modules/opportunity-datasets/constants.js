//
export const OPPORTUNITY_COLOR = ['#b4c248']
