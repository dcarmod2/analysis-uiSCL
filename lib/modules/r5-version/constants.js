// Bucket where R5 JARs are to be found
export const R5_BUCKET = 'https://r5-builds.s3.amazonaws.com'

// Recommended r5 version
export const RECOMMENDED_R5_VERSION = 'v6.0.1'

// R5 v4 has numerous breaking changes in worker/broker communication
export const MINIMUM_R5_VERSION = 'v4.0.0'

// Determine if a particular version is a release version
export const RELEASE_VERSION_REGEX = /^v[0-9]+\.[0-9]+\.[0-9]+$/

// Parse a version into its constituent pieces (major, minor, patch, distance from last release, commit)
export const VERSION_PARSE_REGEX = /^v([0-9]+).([0-9]+).([0-9]+)-?([0-9]+)?-?(.*)?$/
