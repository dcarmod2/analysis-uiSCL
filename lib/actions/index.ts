import {stringify} from 'querystring'
import {createAction} from 'redux-actions'

import {API} from 'lib/constants'
import setURLSearchParameter from 'lib/utils/set-search-parameter'

import fetch from './fetch'

// For storing the query string
export const setQueryString = createAction('set query string')

// Update a search parameters
export const setSearchParameter = (
  params: string | Record<string, string>,
  value?: string
) => {
  if (typeof params === 'string') {
    const opts = {[`${params}`]: value}
    setURLSearchParameter(opts)
    return {
      type: 'set search parameter',
      payload: opts
    }
  } else {
    setURLSearchParameter(params)
    return {
      type: 'set search parameter',
      payload: params
    }
  }
}

export const clearError = createAction('clear error')

// bundle
export const addBundle = createAction('add bundle')
const deleteBundleLocally = createAction('delete bundle')
export const deleteBundle = (_id) =>
  fetch({
    url: `${API.Bundle}/${_id}`,
    options: {
      method: 'delete'
    },
    next: () => deleteBundleLocally(_id)
  })
export const saveBundle = (bundle) =>
  fetch({
    url: `${API.Bundle}/${bundle._id}`,
    options: {
      method: 'put',
      body: bundle
    },
    next: (response) => setBundle(response.value)
  })
export const setBundle = createAction('set bundle')
export const setBundles = createAction('set bundles')
export const loadBundle = (_id) =>
  fetch({
    url: `${API.Bundle}/${_id}`,
    next: (response) => setBundle(response.value)
  })
export const loadBundles = (query) =>
  fetch({
    url: `${API.Bundle}?${stringify(query)}`,
    next: (response) => setBundles(response.value)
  })

// feed
/** Update the data pulled in from the GTFS feed */
export const setFeeds = createAction('set feeds')

// login / logout
export const login = createAction('log in')
export const setUser = createAction('set user')

export default {}
