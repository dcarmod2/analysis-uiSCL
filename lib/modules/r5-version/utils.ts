import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
