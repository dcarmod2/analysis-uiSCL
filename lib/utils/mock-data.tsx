import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'util'
import {constructor as XorShift} from 'xorshift'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN, TRAVEL_TIME_PERCENTILES} from '../constants'

import multi from './multi'

// Create a random number generator that is seeded
const seededRandom = new XorShift([1, 2, 3, 4])

export const makeMockStore = configureStore([multi, thunkMiddleware])

export const mockWithProvider = (children, mockData = mockStores.init) => {
  const store = makeMockStore(mockData)
  const wrapper = mount(<Provider store={store}>{children}</Provider>)

  return {
    snapshot: () => mountToJson(wrapper),
    store,
    wrapper
  }
}

export const mockModel = {
  accessGroup: 'accessGroupName',
  createdAt: '2017-11-08T12:30:50.133Z',
  createdBy: 'person@gmail.com',
  nonce: '12345',
  updatedAt: '2017-11-08T12:31:19.602Z',
  updatedBy: 'person@gmail.com'
}

export const mockSegmentGeom: GeoJSON.LineString = {
  type: 'LineString',
  coordinates: [
    [-122.0246, 36.9707],
    [-122.0279, 37.049],
    [-121.9799, 37.2299],
    [-121.9445, 37.324],
    [-121.936, 37.353],
    [-121.924, 37.365]
  ]
}

export const mockSegment: CL.ModificationSegment = {
  fromStopId: '1',
  toStopId: null,
  geometry: mockSegmentGeom,
  spacing: 3000,
  stopAtEnd: false,
  stopAtStart: false
}

export const mockStop1 = {
  autoCreated: false,
  bearing: 0,
  center: [36.9707, -122.0246],
  distanceFromStart: 0,
  index: 0,
  lat: 36.9707,
  lon: -122.0246,
  stopId: '1'
}

export const mockGtfsStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockGtfsStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockGtfsStops = [mockGtfsStop1, mockGtfsStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  name: 'Mock Pattern',
  pattern_id: 'mock-pattern-id',
  stops: mockGtfsStops,
  trips: [
    {
      duration: 1234,
      end_time: 13579,
      start_time: 12345,
      trip_headsign: 'To Downtown',
      trip_id: 'abcd',
      trip_short_name: 'The Express'
    }
  ]
}

export const mockRoute = {
  label: 'mock-route-label',
  patterns: [mockPattern],
  route_id: 'route1'
}

export const mockFeed = {
  checksum: 'MOCK',
  id: '1',
  routes: [mockRoute],
  stopsById: {
    1: mockGtfsStop1,
    2: mockGtfsStop2
  },
  stops: [mockGtfsStop1, mockGtfsStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  _id: 'timetable-id',
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  dwellTimes: [12],
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd',
  phaseFromTimetable: '',
  phaseFromStop: '',
  phaseAtStop: '',
  phaseSeconds: 300,
  exactTimes: false,
  modificationId: '1234',
  modificationName: 'Test Modification'
}

export const mockOpportunityDataset = {
  sourceName: 'lodes-data-2014',
  name: 'Workers total',
  _id: 'Workers_total'
}

const regionId = '1'

export const mockBundle = {
  ...mockModel,
  errorCode: '',
  name: 'Mock Bundle',
  _id: '1',
  centerLat: 38.90124,
  centerLon: -77.015615,
  feeds: [],
  regionId,
  serviceStart: '2018-09-27',
  serviceEnd: '2018-09-28',
  status: 'DONE',
  totalFeeds: 0
}

export const mockRegion = {
  ...mockModel,
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Region description',
  _id: regionId,
  name: 'Mock Region',
  opportunityDatasets: [mockOpportunityDataset],
  statusCode: 'DONE'
}

export const mockProject = {
  ...mockModel,
  bundleId: '1',
  _id: '1',
  name: 'Mock Project',
  regionId: mockRegion._id,
  variants: ['Default']
}

export const mockModification = {
  ...mockModel,
  bidirectional: false,
  description: 'Mock description',
  entries: [],
  feed: mockFeed.id,
  _id: '1234',
  name: 'Test Modification',
  routes: [mockRoute.route_id],
  projectId: mockProject._id,
  segments: [mockSegment],
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [true]
}

export const mockProfileRequest = {
  accessModes: 'WALK',
  bikeSafe: 1,
  bikeSlope: 1,
  bikeSpeed: 4.166666666666667,
  bikeTime: 1,
  bounds: mockRegion.bounds,
  bundleId: mockBundle._id,
  date: '2016-01-16',
  defaultExpanded: true,
  directModes: 'WALK',
  egressModes: 'WALK',
  fromLat: 12,
  fromLon: 34,
  fromTime: 25200,
  maxBikeTime: 20,
  maxRides: 4,
  maxTripDurationMinutes: 60,
  maxWalkTime: 20,
  monteCarloDraws: 200,
  name: 'Profile Request Test',
  projectId: mockProject._id,
  reachabilityThreshold: 0,
  regionId: mockRegion._id,
  toTime: 32400,
  transitModes: 'BUS,TRAM,RAIL,SUBWAY',
  travelTimePercentile: 50,
  variantIndex: 0,
  walkSpeed: 1.3888888888888888,
  workerVersion: 'v4.0.0'
}

export const mockRegionalAnalyses = [
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 0,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1,
    travelTimePercentiles: [50]
  },
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 1,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 50,
    travelTimePercentiles: [50]
  }
]

const GRID_SIZE = 300
const GRID_MAX = 2000
export const mockGrids = mockRegionalAnalyses.map((ra) => ({
  analysisId: ra._id,
  percentile: 50,
  cutoff: 60,
  zoom: 9,
  width: GRID_SIZE,
  height: GRID_SIZE,
  north: 20024,
  west: 20015,
  data: Array(GRID_SIZE * GRID_SIZE)
    .fill(0)
    .map(() => Math.floor(seededRandom.random() * GRID_MAX)),
  min: 0,
  max: GRID_MAX,
  contains: () => true
}))

export const mockGrid = mockGrids[0]

export const mockStores = {
  init: {
    analysis: {
      copyRequestSettings: true,
      isochroneFetchStatus: false,
      regional: {
        bounds: mockRegion.bounds
      },
      requestsSettings: [mockProfileRequest],
      resultsSettings: [],
      maxTripDurationMinutes: 60,
      travelTimePercentile: 50
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    opportunityDatasets: {
      active: mockOpportunityDataset._id,
      datasets: mockRegion.opportunityDatasets,
      uploadStatuses: []
    },
    queryString: {
      analysisId: mockRegionalAnalyses[0]._id,
      cutoff: 60,
      percentile: 50,
      modificationId: mockModification._id,
      projectId: mockProject._id,
      regionId: mockRegion._id
    },
    region: {
      aggregationAreas: [],
      bundles: [mockBundle],
      regions: [mockRegion]
    },
    regionalAnalyses: {
      analyses: mockRegionalAnalyses,
      grids: mockGrids
    },
    project: {
      activeModificationId: mockModification._id,
      currentProject: mockProject,
      feeds: [mockFeed],
      modifications: [mockModification],
      projects: [mockProject]
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  modificationId: 'Add Trip Pattern',
  title: "Error while applying the modification entitled 'Add Trip Pattern'.",
  messages: ['Headway is not greater than zero.']
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  depth: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

/**
 * This is a function because responseToSurface is destructive and can lead to
 * test leakage if we don't have a brand-new surface each time.
 */
export const createMockSurfaceData = () => {
  const {width, height, depth, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    scenarioApplicationWarnings: [mockScenarioApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize =
    9 * 4 + // header
    width * height * depth * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * depth)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = depth

  // phew, header is done
  for (let z = 0, position = 9; z < depth; z++) {
    let prev = 0
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++, position++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[position] = pixelVal - prev // values are delta-coded within each depth
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(
    buffer,
    buffer.byteLength - metadataEncoded.byteLength,
    metadataEncoded.length
  )
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

/**
 * Return the value stored in sample z of pixel x, y of the mock surface
 */
export function getMockSurfaceValue(x, y, z) {
  // This used to simply return z, 2 * z and 3 * z when we had an array of
  // [5, 10, 15 . . . 95]th percentiles. However, when we switched to
  // requesting specific percentiles that stopped working. This translates the
  // position in the access grid into a percentile, then subtracts five and
  // divides by five to get the position in the old array, so that tests
  // written assuming the old structure continue to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return (2 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  if (x < 2 && y >= 2) return (3 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  return 0
}

const opportunityDatasetGridData = new Int32Array(
  mockSurfaceParameters.width * mockSurfaceParameters.height
)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) {
  opportunityDatasetGridData[i] = 1
}
for (
  let i = 3 * mockSurfaceParameters.width;
  i < opportunityDatasetGridData.length;
  i++
) {
  opportunityDatasetGridData[i] = 2
}

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: opportunityDatasetGridData,
  getValue(x, y) {
    return opportunityDatasetGridData[y * mockSurfaceParameters.width + x]
  }
}

export const releaseVersions = ['v3.0.0', 'v2.2.0', 'v2.0.0']
export const allVersions = [
  'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0'
]
