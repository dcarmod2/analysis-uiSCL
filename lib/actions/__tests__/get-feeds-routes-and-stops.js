import nock from 'nock'

import {makeMockStore, mockModification} from 'lib/utils/mock-data'
import {REROUTE} from 'lib/constants'

import getFeedsRoutesAndStops from '../get-feeds-routes-and-stops'

describe('actions > getFeedsRoutesAndStops', () => {
  const mockRoutes = [
    {
      route_id: '1',
      route_short_name: '1'
    },
    {
      route_id: '2',
      route_short_name: '2'
    }
  ]

  const mockResponse = {
    bundle: [
      {
        feeds: [
          {
            checksum: 'abcd',
            detailRoutes: mockRoutes,
            feed_id: '1',
            routes: mockRoutes,
            stops: [
              {
                stop_id: '1'
              }
            ]
          }
        ]
      }
    ]
  }

  // add a mock unfetched route so there is a need to fetch
  const mockModificationWithUnfetchedRoute = {
    ...mockModification,
    type: REROUTE
  }

  const unfetchedActionParams = {
    bundleId: '1',
    modifications: [mockModificationWithUnfetchedRoute]
  }

  const mockHeaders = {
    Content: 'json'
  }

  it('should work when fetching via getAllRoutesAndStops', () => {
    nock('http://localhost')
      .get(/^\/api\/graphql/)
      .reply(200, mockResponse, mockHeaders)

    const store = makeMockStore()
    return store
      .dispatch(
        getFeedsRoutesAndStops({
          bundleId: '1',
          forceCompleteUpdate: true
        })
      )
      .then((r) => {
        expect(r).toMatchSnapshot()
      })
  })

  it('should work when fetching with getUnfetchedRoutes', () => {
    // make request again, forcing fetching via getUnfetchedRoutes
    nock('http://localhost')
      .get(/^\/api\/graphql/)
      .reply(200, mockResponse, mockHeaders)

    const store = makeMockStore()
    return store
      .dispatch(getFeedsRoutesAndStops(unfetchedActionParams))
      .then((r) => {
        expect(r).toMatchSnapshot()
      })
  })

  it('should fetch new routes', () => {
    // make request again, forcing fetching via getUnfetchedRoutes
    nock('http://localhost')
      .get(/^\/api\/graphql/)
      .reply(200, {bundle: []}, mockHeaders)

    const store = makeMockStore()
    const get = getFeedsRoutesAndStops(unfetchedActionParams)
    return store.dispatch(get).then(() => {
      expect(store.getActions()).toMatchSnapshot()
    })
  })
})
