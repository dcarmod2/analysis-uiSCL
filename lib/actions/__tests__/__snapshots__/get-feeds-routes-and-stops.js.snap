// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions > getFeedsRoutesAndStops should fetch new routes 1`] = `
Array [
  Object {
    "payload": Object {
      "id": 3,
      "options": Object {},
      "type": "__FETCH__",
      "url": "http://localhost/api/graphql?query=%0Aquery%20dataQuery(%24bundleId%3A%20String%2C%20%24routeIds%3A%20%5BString%5D)%20%7B%0A%20%20bundle%20(bundle_id%3A%20%5B%24bundleId%5D)%20%7B%0A%20%20%20%20feeds%20%7B%0A%20%20%20%20%20%20feed_id%2C%0A%20%20%20%20%20%20checksum%2C%0A%20%20%20%20%20%20routes%20%7B%0A%20%20%20%20%20%20%20%20route_id%0A%20%20%20%20%20%20%20%20route_short_name%0A%20%20%20%20%20%20%20%20route_long_name%0A%20%20%20%20%20%20%7D%0A%0A%20%20%20%20%20%20detailRoutes%3A%20routes(route_id%3A%20%24routeIds)%20%7B%0A%20%20%20%20%20%20%20%20route_id%0A%20%20%20%20%20%20%20%20route_short_name%0A%20%20%20%20%20%20%20%20route_long_name%0A%20%20%20%20%20%20%20%20patterns%20%7B%0A%20%20%20%20%20%20%20%20%20%20name%2C%0A%20%20%20%20%20%20%20%20%20%20pattern_id%2C%0A%20%20%20%20%20%20%20%20%20%20geometry%2C%0A%20%20%20%20%20%20%20%20%20%20trips%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20trip_id%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20trip_short_name%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20trip_headsign%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20start_time%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20direction_id%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20duration%0A%20%20%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%20%20%20%20%20%20stops%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20stop_id%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%0A%20%20%20%20%20%20stops%20%7B%0A%20%20%20%20%20%20%20%20stop_id%2C%0A%20%20%20%20%20%20%20%20stop_name%2C%0A%20%20%20%20%20%20%20%20stop_lat%2C%0A%20%20%20%20%20%20%20%20stop_lon%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A&variables=%7B%22bundleId%22%3A%221%22%2C%22routeIds%22%3A%5B%22route1%22%5D%7D",
    },
    "type": "increment outstanding fetches",
  },
  Object {
    "payload": Object {
      "id": 3,
      "type": "__FETCH__",
    },
    "type": "decrement outstanding fetches",
  },
  Object {
    "payload": Array [],
    "type": "set feeds",
  },
]
`;

exports[`actions > getFeedsRoutesAndStops should work when fetching via getAllRoutesAndStops 1`] = `
Array [
  Object {
    "checksum": "abcd",
    "id": "1",
    "routes": Array [
      Object {
        "label": "1 ",
        "route_id": "1",
        "route_short_name": "1",
      },
      Object {
        "label": "2 ",
        "route_id": "2",
        "route_short_name": "2",
      },
    ],
    "stops": Array [
      Object {
        "stop_id": "1",
      },
    ],
    "stopsById": Object {
      "1": Object {
        "stop_id": "1",
      },
    },
  },
]
`;

exports[`actions > getFeedsRoutesAndStops should work when fetching with getUnfetchedRoutes 1`] = `
Array [
  Object {
    "checksum": "abcd",
    "id": "1",
    "routes": Array [
      Object {
        "label": "1 ",
        "route_id": "1",
        "route_short_name": "1",
      },
      Object {
        "label": "2 ",
        "route_id": "2",
        "route_short_name": "2",
      },
    ],
    "stops": Array [
      Object {
        "stop_id": "1",
      },
    ],
    "stopsById": Object {
      "1": Object {
        "stop_id": "1",
      },
    },
  },
]
`;
